Plug-Ins/NovaTools/System Editor/StellarView.swift

```diff
@@ -0,0 +1,210 @@
+        } catch {
+            return nil
+        }
+    }
+
+    func read() throws {
+        let reader = BinaryDataReader(resource.data)
+        position.x = Double(try reader.read() as Int16)
+        position.y = Double(try reader.read() as Int16)
+
+        let spinId = (Int)(try reader.read() as Int16) + 1000;
+        if let spinResource = manager.findResource(type: ResourceType("spïn"), id: spinId, currentDocumentOnly: false) {
+            let spinReader = BinaryDataReader(spinResource.data)
+            let spriteId = (Int)(try spinReader.read() as Int16);
+
+            // Use the preview provider to obtain an NSImage from the resources the spïn might point to -- try PICT then fall back to rlëD
+            if let spriteResource = manager.findResource(type: ResourceType("PICT"), id: spriteId, currentDocumentOnly: false) ?? manager.findResource(type: ResourceType("rlëD"), id: spriteId, currentDocumentOnly: false) {
+                spriteResource.preview({ img in
+                    self.image = img
+                    self.updateFrame()
+                    self.needsDisplay = true
+                })
+            }
+        }
+    }
+
+    override func updateTrackingAreas() {
+        for trackingArea in self.trackingAreas {
+            self.removeTrackingArea(trackingArea)
+        }
+        let tracking = NSTrackingArea(rect: bounds, options: [.mouseEnteredAndExited, .activeInKeyWindow, .enabledDuringMouseDrag], owner: self, userInfo: nil)
+        self.addTrackingArea(tracking)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func updateFrame() {
+        if image != nil, let imageSize = systemView?.transform.transform(image!.size) {
+            frame.size.width = imageSize.width
+            frame.size.height = imageSize.height
+        } else {
+            frame.size.width = 12
+            frame.size.height = 12
+        }
+
+        if let point = systemView?.transform.transform(position) {
+            self.point = point
+        }
+    }
+
+    func move(to newPos: NSPoint) {
+        let curPos = position
+        position = NSPoint(x: newPos.x.rounded(), y: newPos.y.rounded())
+        self.updateFrame()
+        self.save()
+        undoManager?.registerUndo(withTarget: self) { $0.move(to: curPos) }
+    }
+
+    private func save() {
+        guard resource.data.count >= 2 * 2 else {
+            return
+        }
+        let writer = BinaryDataWriter()
+        writer.data = resource.data
+        writer.write(Int16(position.x), at: 0)
+        writer.write(Int16(position.y), at: 2)
+        systemView?.isSavingStellar = true
+        resource.data = writer.data
+        systemView?.isSavingStellar = false
+        systemView?.needsDisplay = true
+    }
+
+    override func draw(_ dirtyRect: NSRect) {
+        guard dirtyRect.intersects(bounds) else {
+            return
+        }
+
+        if image != nil {
```

Comment:
Also here, so you can avoid the forced unwraps later.
```suggestion
        if let image {
```